### PR TITLE
util: Add support for more than two cliques

### PR DESCRIPTION
**Pull request by wesbland**
The `MPIR_CVAR_ODD_EVEN_CLIQUES` variable only supports two cliques per
node. By creating a new CVAR that accepts an integer instead of boolean
value, we can create as many local cliques as we want (up to the support
level of the machine).

Deprecate `MPIR_CVAR_ODD_EVEN_CLIQUES` in favor of the new
`MPIR_CVAR_NUM_CLIQUES`.

### CODE REVIEW

File: src/util/nodemap/build_nodemap.h [hunk 1]
```diff
@@ -497,9 +510,15 @@ static inline int MPIR_NODEMAP_build_nodemap(int sz,
...
     }
 
-odd_even_cliques:
-    if (odd_even_cliques)
-    {
+cliques:
+    if (MPIR_CVAR_NUM_CLIQUES > 1) {
+        for (i = 0; i < sz; ++i) {
+            if (i % MPIR_CVAR_NUM_CLIQUES) {
+                out_nodemap[i] += g_max_node_id + (i % MPIR_CVAR_NUM_CLIQUES );
```
Comment (wesbland) [inline src/util/nodemap/build_nodemap.h]: I see what you're saying. That wasn't my intention to have those overlapping. I'll have to fix that.

**Comment by wesbland** [comment]
The use case is that sometimes it's nice to be able to simulate more nodes than just two on a single node.

My particular motivation was to be able to determine memory usage on a large cluster than I had available at the time. I could put every MPI process on its own node and simulate hundreds of nodes with just a few physical machines.

**Comment by yfguo** [comment]
I see. That would be useful.

**Comment by wesbland** [comment]
Ok @yfguo, I think this is correct now. Can you take another look?

File: src/util/nodemap/build_nodemap.h [hunk 2]
```diff
@@ -497,9 +510,15 @@ static inline int MPIR_NODEMAP_build_nodemap(int sz,
...
-    if (odd_even_cliques)
-    {
+cliques:
+    if (MPIR_CVAR_NUM_CLIQUES > 1) {
+        for (i = 0; i < sz; ++i) {
+            if (i % MPIR_CVAR_NUM_CLIQUES) {
+                out_nodemap[i] += (g_max_node_id + 1) * (i % MPIR_CVAR_NUM_CLIQUES);
+            }
+        }
+        g_max_node_id = (g_max_node_id + 1) * MPIR_CVAR_NUM_CLIQUES;
```
Comment (yfguo) [inline src/util/nodemap/build_nodemap.h]: Can you change this line to `g_max_node_id=(g_max_node_id + 1) * MPIR_CVAR_NUM_CLIQUES - 1;`? The "build_nodemap" function gives "maximum node id" in `g_max_node_id`, but after this code `g_max_node_id` will essentially be "number of nodes".

The current MPICH has some inconsistency (in CH3 and CH4) on whether `g_max_node_id` is "maximum node ID" or "number of nodes". Ken and I discussed this and we think we should clarify it as "maximum node ID", and I have a PR pending for it. So, I think it is better to get it right for this patch.

**Comment by yfguo** [comment]
On a side note, we should also consider adding some unittests after this patch goes in and we have some experience with multiclique.

File: src/util/nodemap/build_nodemap.h (hunk 2 above)
Comment (wesbland) [inline src/util/nodemap/build_nodemap.h]: Done

**Comment by wesbland** [comment]
Doh! Missed parens. 😢 

Feel free to squash those two commits together.

**Comment by yfguo** [comment]
Good to go.

**Pull request merged**